### PR TITLE
Set-up Auth0 secure route on back-end

### LINKED ISSUE #26: Integrate Auth with backend
- update db 
- register app as authenticated using Google console

### PR DESCRIPTION
**Pull request by armiantos**
# Overview

* Set-up secure routes using Auth0
* `api/secure/v1` routes require an API token requested by the client with an audience (or API identifier) and scope defined by us in Auth0 API section
* Give sample secure API call

# Changes

* Install `cors` server side (I'm not sure if this is safe to keep)
* Create jwt checker middleware to ensure client authenticity on secure routes
* Install `axios` on React client to make server requests
* Set-up environment variables for Auth0 variables

# Questions & Notes

Are there any open questions? Anything else important we should know?

# Issue tracking

Closes #26 

# Testing & Demo

Try `curl`ing the secure ping endpoint without a token

image

Pressing the button in the React app after signed in should give you the pong response in the browser console

image

# TODO

* Write tests for middleware

### CODE REVIEW

File: api/src/routes/middleware.ts
Comment (calebschoepp) [reply api/src/routes/middleware.ts:35]: Another thing to note is that I would like to follow the pattern I outline in #56 with the `authorize` and `authenticate` endpoints. That is, if they fail they throw the appropriate `AuthenticationException` or `AuthorizationException` which is caught by the error handling middleware which will return the appropriate response.

Don't worry about throwing the exceptions in this PR though. #56 needs to merge first.

File: www/src/util/config.ts
Comment (calebschoepp) [inline www/src/util/config.ts:5]: IMO these should have sane defaults for local development e.g. `const SERVER_ENDPOINT = process.env.REACT_APP_SERVER_ENDPOINT || 'http://localhost:1337';`

File: www/src/providers/Auth0ProviderWithHistory.tsx
Comment (calebschoepp) [inline www/src/providers/Auth0ProviderWithHistory.tsx:20]: Fine for now, but those scopes should really become a TypeScript Enum.

Also question, eventually will the scopes this component needs to claim basically just be all possible scopes?

File: www/src/routes/ResumeReview.tsx
Comment (calebschoepp) [inline www/src/routes/ResumeReview.tsx:14]: What happens if the user doesn't have the `call:ping` scope? Will it just get an access token but get `403`'s at the server. OR will it fail to get access token from Auth0?

File: www/src/routes/ResumeReview.tsx
Comment (calebschoepp) [inline www/src/routes/ResumeReview.tsx:9]: Maybe stick a `// TODO: Remove this` above this so we don't forget to take a random call to our api out of the code when we ship this.

**Review by calebschoepp — CHANGES_REQUESTED** [review]
Amazing. Auth is looking easier and easier. Thanks for this work. Left a bunch of comments I'd like resolved before the approval but we are real close.

File: api/package.json
Comment (armiantos) [reply api/package.json:42]: Opened #65

File: api/package.json
Comment (armiantos) [reply api/package.json:48]: There's no need to add `@types` on these, 2/3 of them already support native Typescript, namely

image
image

Do we want to explicitly add the type dependency for `express-jwt`? `jwks-rsa` already adds this dependency to us

File: api/src/routes/middleware.ts
Comment (armiantos) [reply api/src/routes/middleware.ts]: Sounds like a plan! Opened #66

File: api/src/routes/middleware.ts
Comment (armiantos) [reply api/src/routes/middleware.ts:39]: I did it because `jwtCheck` needs to have a next function

File: api/src/server.ts
Comment (armiantos) [reply api/src/server.ts:24]: Agreed

File: api/src/util/jwtCheck.ts
Comment (armiantos) [reply api/src/util/jwtCheck.ts]: Good catch! Will update

File: www/src/providers/Auth0ProviderWithHistory.tsx
Comment (armiantos) [reply www/src/providers/Auth0ProviderWithHistory.tsx:20]: Yes that's what I imagined, the returned scopes is an intersection of the scopes here and what the user actually has. If the user doesn't specify any scope in the beginning, we'll need to spawn another login or consent page to re-fetch the scope in the token

File: www/src/routes/ResumeReview.tsx
Comment (armiantos) [reply www/src/routes/ResumeReview.tsx:9]: I think it might be best to just remove it entirely and cross this bridge again when we need to. This was just for the purposes of the demo.

File: www/src/routes/ResumeReview.tsx
Comment (armiantos) [reply www/src/routes/ResumeReview.tsx:14]: One of two things can happen. 

* If the user is assigned a role with the specified permissions but doesn't have it in the initial token, the error will say "Consent required"
* If the user doesn't have permissions, it'll say insufficient scope

File: api/.env.sample
Comment (armiantos) [reply api/.env.sample]: Sounds good, will add 👍

File: www/src/routes/ResumeReview.tsx
Comment (armiantos) [reply www/src/routes/ResumeReview.tsx]: I'm hoping it throws a differrent exception so that we can display a different message for each

File: api/package.json
Comment (calebschoepp) [reply api/package.json:48]: IMO better to just be explicit with it. Probably never going to happen but you could imagine a future in which `jwks-rsa` no longer provides the `@types` dependency for us.

File: www/src/providers/Auth0ProviderWithHistory.tsx
Comment (calebschoepp) [reply www/src/providers/Auth0ProviderWithHistory.tsx:20]: Sounds good.

File: www/src/routes/ResumeReview.tsx
Comment (calebschoepp) [reply www/src/routes/ResumeReview.tsx]: Okay, and so it seems that this will all happen client side i.e. you don't need to hit the api with invalid token to learn that it is invalid.

File: www/src/routes/ResumeReview.tsx
Comment (calebschoepp) [reply www/src/routes/ResumeReview.tsx]: Yep, let's remove it then.

File: api/src/routes/middleware.test.ts
Comment (calebschoepp) [inline api/src/routes/middleware.test.ts:4]: Nit: Verb and noun are still flipped throughout these tests. `jwtCheck` instead of `checkJwt.

File: api/src/routes/middleware.ts
Comment (calebschoepp) [inline api/src/routes/middleware.ts:32]: Nit: No space in `middle ware`

**Review by calebschoepp — APPROVED** [review]
Couple nits but LGTM!

Ping me if you need re-approval when you push changes.

File: api/src/routes/middleware.ts
Comment (armiantos) [reply api/src/routes/middleware.ts:32]: Good catch (2) 😂

**Review by calebschoepp — APPROVED** [review]

**Pull request merged**
